### PR TITLE
Support asynchronous service callbacks

### PR DESCRIPTION
**Pull request by chris-smith**
-Service callbacks can now return promises. Result
 will be sent when promise resolves.
-Adds tests for asynchronous services.
-Service clients whose connections are closed while a call
 is in progress will be rejected.

### CODE REVIEW

**Review by IanTheEngineer — APPROVED** [review]
Looks great! I think this should be good to merge after a rebase

**Comment by blakelyc** [comment]
Correct me if I am wrong, but this why Actions exist.  Services are supposed to be synchronous (blocking).

**Comment by chris-smith** [comment]
TLDR: I think enforcing a synchronous Service callback pattern breaks down a bit in Node.

In this patch we're not stopping people from writing simple, synchronous callback functions that return immediately - if their callback function returns a boolean response we'll consider the service complete. I do think it'd be strange if users returned `true` in their callback but also kicked off some asynchronous work.

But if in order to do the service's job it has to do some asynchronous work (which is very common in Node) then I think we should let them and should delay sending the response until they're done. These services don't necessarily need to support preemption or provide progress, but they may still need to be asynchronous because of the nature of the language and the context in which the service is being provided.

Blocking service callbacks in Node.js will also forcibly block anything else from happening (subscription callbacks, web server responses, etc.) which isn't necessarily the case in other implementations (`roscpp` or `rospy`) where that work can be done in separate threads. Actions may be a bit heavy-handed for a lot of these services - I'm assuming actions require more overhead as well because of their complexity.

I'd also take the position that Actions exist to support multiple long-running goals, provide progress on those goals, and allow them to be canceled or interrupted cleanly while Services exist to provide a simple call-and-response mechanism. I've always assumed that the synchronous-Service-callback-pattern arose because callback function signatures implied so in dominant implementations - and because actions came along that made it unnecessary to implement services any differently for those languages.

All that being said, I'm open to discussing this with other people in the Node/ROS community.

**Comment by NickZ** [comment]
Synchronous service callbacks are the bane of my existence, and have caused us so many problems in trying to support services in our software. I would love love love for this to be merged in.

**Comment by chris-smith** [comment]
Barring some negative response to this, I'll likely merge in the next couple of days.
